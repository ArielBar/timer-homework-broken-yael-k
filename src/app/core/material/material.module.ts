import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatIconModule } from '@angular/material/icon';
import { MatCardModule } from '@angular/material/card';
import { MatButtonModule } from '@angular/material/button';
import { MatInputModule } from '@angular/material/input';

const modules = [
  CommonModule,
  MatCardModule,
  MatButtonModule,
  MatInputModule,
  MatIconModule,
];
@NgModule({
  imports: [...modules],
  exports: [...modules],
})
export class MaterialModule {}
