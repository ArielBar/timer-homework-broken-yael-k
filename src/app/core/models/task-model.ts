import { Observable } from 'rxjs';
import { buttonText } from './button-text';

export interface TaskModel {
  id: number;
  name: string;
  timer: Observable<number>;
  buttonText: buttonText;
}
