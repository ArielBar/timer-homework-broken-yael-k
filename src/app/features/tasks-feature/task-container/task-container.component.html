<mat-card>
  <app-task-presenter
    *ngFor="let task of tasks$ | async"
    [name]="task.name"
    [id]="task.id"
    [timer]="task.timer"
    [buttonText]="task.buttonText"
    (clicked)="onClick($event)"
  ></app-task-presenter>
</mat-card>
<mat-card>
  <mat-card-title>
    Total: {{ totalTime$ | async | minuteSeconds }}
  </mat-card-title>
</mat-card>
