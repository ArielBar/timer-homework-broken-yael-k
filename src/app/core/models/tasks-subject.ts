import { BehaviorSubject } from 'rxjs';
import { TaskModel } from './task-model';
export class TasksSubject extends BehaviorSubject<TaskModel[]> {
  constructor(state: TaskModel[]) {
    super(state || []);
  }

  addTask(task: TaskModel): void {
    const tasks = this.getValue();
    tasks.push(task);
    this.next(tasks);
  }

  removeTask(task: TaskModel): void {
    const tasks = this.getValue();
    const index = tasks.indexOf(task);
    if (index !== -1) {
      tasks.splice(index, 1);
      this.next(tasks);
    }
  }

  updateTask(task: TaskModel): void {
    const tasks = this.getValue();
    const index = tasks.findIndex((t) => t.id === task.id);
    if (index !== -1) {
      tasks[index] = task;
      this.next(tasks);
    }
  }

  setTasks(tasks: TaskModel[]) {
    this.next(tasks);
  }
}
