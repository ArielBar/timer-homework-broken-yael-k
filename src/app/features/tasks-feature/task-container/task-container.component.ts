import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
} from '@angular/core';
import { Observable } from 'rxjs';
import { TaskModel } from '../../../core/models/task-model';
import { TaskStoreService } from '../../../core/services/task-store.service';

@Component({
  selector: 'app-task-container',
  templateUrl: './task-container.component.html',
  styleUrls: ['./task-container.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TaskContainerComponent implements OnInit {
  public tasks$: Observable<TaskModel[]>;
  public totalTime$: Observable<number>;
  constructor(private service: TaskStoreService) {}

  ngOnInit() {
    this.tasks$ = this.service.tasks$;
    this.totalTime$ = this.service.totalTime$;
  }

  public onClick(evt: number) {
    this.service.updateTask(evt);
  }
}
