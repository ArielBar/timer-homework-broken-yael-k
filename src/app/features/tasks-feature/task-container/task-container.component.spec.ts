import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Input, Component } from '@angular/core';
import { TaskModel } from '../../../core/models/task-model';
import { TaskStoreService } from '../../../core/services/task-store.service';
import { TaskContainerComponent } from './task-container.component';
import { MatCardModule } from '@angular/material/card';
import { MinuteSecondsPipe } from '../../../core/pipes/minute-seconds.pipe';
import { Observable, of } from 'rxjs';
import { buttonText } from 'src/app/core/models/button-text';

describe('TaskContainerComponent', () => {
  let component: TaskContainerComponent;
  let fixture: ComponentFixture<TaskContainerComponent>;

  @Component({
    selector: 'app-task-presenter',
    template: ` <p>{{ task | json }}</p> `,
  })
  class FakeComponent implements TaskModel {
    @Input() id: number;
    @Input() name: string;
    @Input() timer: Observable<number>;
    @Input() buttonText: buttonText;
  }

  class FakeService {
    get tasks$() {
      return of([]);
    }
    get totalTime$() {
      return of(1);
    }
    updateTask() {
      return {};
    }
  }
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TaskContainerComponent, FakeComponent, MinuteSecondsPipe],
      providers: [{ provide: TaskStoreService, useClass: FakeService }],
      imports: [MatCardModule],
    });
    fixture = TestBed.createComponent(TaskContainerComponent);
    component = fixture.componentInstance;
  });
  it('can load instance', () => {
    expect(component).toBeTruthy();
  });
  describe('onClick', () => {
    it('makes expected calls', () => {
      const taskModelStub: TaskModel = {
        id: -1,
        name: '',
        timer: of(),
        buttonText: 'play_arrow',
      };
      const logicServiceStub: TaskStoreService =
        fixture.debugElement.injector.get(TaskStoreService);
      jest.spyOn(logicServiceStub, 'updateTask');
      component.onClick(-1);
      expect(logicServiceStub.updateTask).toHaveBeenCalled();
    });
  });
  describe('DOM', () => {
    it('should render totalTime$', () => {
      const logicServiceStub: TaskStoreService =
        fixture.debugElement.injector.get(TaskStoreService);
      jest.spyOn(logicServiceStub, 'totalTime$', 'get').mockReturnValue(of(35));
      fixture.detectChanges();

      fixture.whenStable().then(() => {
        expect(fixture.nativeElement).toMatchSnapshot();
      });
    });
    it('should render tasks$', () => {
      const logicServiceStub: TaskStoreService =
        fixture.debugElement.injector.get(TaskStoreService);
      const tasks: TaskModel[] = [
        { id: 1, name: 'test1', buttonText: 'pause', timer: of(1) },
        { id: 2, name: 'test2', buttonText: 'pause', timer: of(1) },
        { id: 3, name: 'test3', buttonText: 'pause', timer: of(1) },
      ];
      jest.spyOn(logicServiceStub, 'tasks$', 'get').mockReturnValue(of(tasks));
      fixture.detectChanges();

      fixture.whenStable().then(() => {
        expect(fixture.nativeElement).toMatchSnapshot();
      });
    });
  });
});
