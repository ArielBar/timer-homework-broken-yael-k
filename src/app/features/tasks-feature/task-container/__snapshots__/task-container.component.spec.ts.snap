// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskContainerComponent DOM should render tasks$ 1`] = `
<div
  id="root3"
>
  <mat-card
    class="mat-card mat-focus-indicator"
  >
    <app-task-presenter>
      <p>
        
      </p>
    </app-task-presenter>
    <app-task-presenter>
      <p>
        
      </p>
    </app-task-presenter>
    <app-task-presenter>
      <p>
        
      </p>
    </app-task-presenter>
    
  </mat-card>
  <mat-card
    class="mat-card mat-focus-indicator"
  >
    <mat-card-title
      class="mat-card-title"
    >
       Total: 00:01 
    </mat-card-title>
  </mat-card>
</div>
`;

exports[`TaskContainerComponent DOM should render totalTime$ 1`] = `
<div
  id="root2"
>
  <mat-card
    class="mat-card mat-focus-indicator"
  >
    
  </mat-card>
  <mat-card
    class="mat-card mat-focus-indicator"
  >
    <mat-card-title
      class="mat-card-title"
    >
       Total: 00:35 
    </mat-card-title>
  </mat-card>
</div>
`;
